packages/eslint-plugin/docs/rules/typedef.md

```diff
@@ -0,0 +1,364 @@
+
+```ts
+class ContainsText {
+  // There must be a type annotation here to infer the type
+  delayedText: string;
+
+  // `typedef` requires a type annotation here to maintain consistency
+  immediateTextExplicit: string = 'text';
+
+  // This is still a string type because of its initial value
+  immediateTextImplicit = 'text';
+}
+```
+
+> Note: requiring type annotations unnecessarily can be cumbersome to maintain and generally reduces code readability.
+> TypeScript is often better at inferring types than easily written type annotations would allow.
+> Instead of enabling `typedef`, it is generally recommended to use the `--noImplicitAny` and/or `--strictPropertyInitialization` compiler options to enforce type annotations only when useful.
+
+## Rule Details
+
+This rule can enforce type annotations in locations regardless of whether they're required.
+This is typically used to maintain consistency for element types that sometimes require them.
+
+## Options
+
+This rule has an object option that may receive any of the following as booleans:
+
+- `"arrayDestructuring"`
+- `"arrowCallSignature"`: `true` by default
+- `"arrowParameter"`: `true` by default
+- `"callSignature"`: `true` by default
+- `"memberVariableDeclaration"`: `true` by default
+- `"objectDestructuring"`
+- `"parameter"`: `true` by default
+- `"propertyDeclaration"`: `true` by default
+- `"variableDeclaration"`
+
+For example, with the following configuration:
+
+```json
+{
+  "rules": {
+    "typedef": [
+      "error",
+      {
+        "arrowParameter": false,
+        "variableDeclaration": true
+      }
+    ]
+  }
+}
+```
+
+- Type annotations on arrow function parameters are not required
+- Type annotations on variables are required
+- Options otherwise adhere to the defaults
+
+### arrayDestructuring
+
+Whether to enforce type annotations on variables declared using array destructuring.
+
+Examples of **incorrect** code with `{ "arrayDestructuring": true }`:
+
+```ts
+const [a] = [1];
+const [b, c] = [1, 2];
+```
+
+Examples of **correct** code with `{ "arrayDestructuring": true }`:
+
+```ts
+const [a]: number[] = [1];
+const [b]: [number] = [2];
+const [c, d]: [boolean, string] = [true, 'text'];
+```
+
+### arrowCallSignature
+
+Whether to enforce type annotations for return types of arrow functions.
+
+Examples of **incorrect** code with `{ "arrowCallSignature": true }`:
+
+```ts
+const returnsVoid = () => {};
+const returnsString = () => 'text';
+
+['hello', 'world'].map(text => text.length);
+
+const mapper = {
+  map: text => text + '...',
+};
+```
+
+Examples of **correct** code with `{ "arrowCallSignature": true }`:
+
+```ts
+const returnsVoid = (): void => {};
+const returnsString = (): string => 'text';
+
+['hello', 'world'].map((text): number => text.length);
+
+const mapper = {
+  map: (text): string => text + '...',
+};
+```
+
+### arrowParameter
+
+Whether to enforce type annotations for parameters of arrow functions.
```

Comment:
isn't this covered by the compiler option `noImplicitAny`?

Reply:
Yeah. Some users use the rule to make it easier to onboard `noImplicitAny`, or as a stopgap until they get around to enabling it.